diff --git a/minisat/utils/System.cc b/minisat/utils/System.cc
index 6d7b684..b9b55df 100644
--- a/minisat/utils/System.cc
+++ b/minisat/utils/System.cc
@@ -97,10 +97,13 @@ double Minisat::memUsedPeak() { return 0; }
 
 void Minisat::setX86FPUPrecision()
 {
-#if defined(__linux__) && defined(_FPU_EXTENDED) && defined(_FPU_DOUBLE) && defined(_FPU_GETCW)
-    // Only correct FPU precision on Linux architectures that needs and supports it:
-    fpu_control_t oldcw, newcw;
-    _FPU_GETCW(oldcw); newcw = (oldcw & ~_FPU_EXTENDED) | _FPU_DOUBLE; _FPU_SETCW(newcw);
+#if defined(__x86_64) || defined(__i386) || defined(_M_IX86) || defined(_X86_) || defined(__X86__)
+    // Only correct FPU precision on architectures that need and support it:
+    fenv_t fenv;
+    fegetenv(&fenv);
+    fenv.__control_word &= ~0x300; // _FPU_EXTENDED
+    fenv.__control_word |= 0x200; // _FPU_DOUBLE
+    fesetenv(&fenv);
     printf("WARNING: for repeatability, setting FPU to use double precision\n");
 #endif
 }
diff --git a/minisat/utils/System.h b/minisat/utils/System.h
index a51d4c2..1e1b5bd 100644
--- a/minisat/utils/System.h
+++ b/minisat/utils/System.h
@@ -21,9 +21,7 @@ OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWA
 #ifndef Minisat_System_h
 #define Minisat_System_h
 
-#if defined(__linux__)
-#include <fpu_control.h>
-#endif
+#include <fenv.h>
 
 #include "minisat/mtl/IntTypes.h"
 
